aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -0,0 +1,202 @@
+ * native fun create_aggregator<Element>(limit: Element): Aggregator<Element>;
+ **************************************************************************************************/
+
+fn native_create_aggregator(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_CREATE_AGGREGATOR_BASE)?;
+    // Get the current aggregator data.
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+    let id = AggregatorID::ephemeral(aggregator_data.generate_id());
+
+    match ty_args[0] {
+        Type::U128 => {
+            let limit = safely_pop_arg!(args, u128);
+            aggregator_data.create_new_aggregator(id, limit);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![
+                Value::u128(0),
+                Value::u128(limit),
+            ]))])
+        },
+        Type::U64 => {
+            let limit = safely_pop_arg!(args, u64);
+            aggregator_data.create_new_aggregator(id, limit as u128);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![
+                Value::u64(0),
+                Value::u64(limit),
+            ]))])
+        },
+        _ => Err(PartialVMError::new(StatusCode::ABORTED)
+            .with_message("Unsupported type supplied to aggregator".to_string())
+            .with_sub_status(0x02_0005)
+            .into()),
+    }
+}
+
+/***************************************************************************************************
+ * native fun try_add<Element>(aggregator: &mut Aggregator<Element>, value: Element): bool;
+ **************************************************************************************************/
+fn native_try_add(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 2);
+
+    context.charge(AGGREGATOR_V2_TRY_ADD_BASE)?;
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    match ty_args[0] {
+        Type::U128 => {
+            // Get aggregator information and a value to add.
+            let value = safely_pop_arg!(args, u128);
+            let (id, limit) = aggregator_info_u128(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit)?;
+            Ok(smallvec![Value::bool(aggregator.try_add(value).is_ok())])
+        },
+        Type::U64 => {
+            // Get aggregator information and a value to add.
+            let value = safely_pop_arg!(args, u64);
+            let (id, limit) = aggregator_info_u64(&safely_pop_arg!(args, StructRef))?;
```

Comment:
If we return limit as u128, we can avoid casts, then we can factor out the code a lot, e.g. using a simple macro. Cause this seems identical?

Reply:
Aggregator is now a generic struct as well, and limit is also u64 inside. 

that makes the API to creating snapshots cleaner, and the only place we need to deal with casting is here. 

this pattern repeats below as well, we can either do a match to just give us a limit in u128:

```
let (value, limit) = match ty_args[0] { 
   ...
};
let aggregator = aggregator_data.get_aggregator(id, limit)?;
Ok(smallvec![Value::bool(aggregator.try_add(value).is_ok())])
```

that might be a bit cleaner.

But as it is repeated across functions, maybe we should just have utility functions pop_arg_into_u128 and pop_arg_aggregator_into_limit_u128:

```
let value = pop_arg_into_u128(args, ty_args[0]);
let limit = pop_arg_aggregator_into_limit_u128(args, ty_args[0]);
let aggregator = aggregator_data.get_aggregator(id, limit)?;
Ok(smallvec![Value::bool(aggregator.try_add(value).is_ok())])

```

Reply:
I tried to used your ideas and abstract out the common code in try_add, try_sub and put it in a separate function. Please let me know if this looks good.